src/deco/deco.hpp

```diff
@@ -213,66 +212,120 @@ class ViewDecoration : public wf::compositor_surface_t,
+
+    /// Surface swap, used when hiding the surface from the node.
+    std::unique_ptr<wf::surface_interface_t> surface;
+
+    /// The loaded options from the cfg.
+    nonstd::observer_ptr<Options> options;
 
     wf::signal_connection_t on_prefered_split_type_changed =
         [&](wf::signal_data_t *) { damage(); };
 
     wf::signal_connection_t on_config_changed = [&](wf::signal_data_t *) {
         // Refresh geometry in case border_width changes.
         node->refresh_geometry();
-
-        // Refreshing the geometry may not actually change the geometry. (e.g.
-        // If only border_radius changes) So we still need to update the
-        // cached_region here.
-        cached_region = calculate_region();
+        surface_ref->recalculate_region();
         node->view->damage();
     };
 
     wf::signal_connection_t on_detached = [&](wf::signal_data_t *) {
-        mapped = false;
-        wf::emit_map_state_change(this);
-        node->view->set_decoration(nullptr); // ViewDecoration dies here.
+        surface_ref->unmap();
+        detach_surface();
+
+        // Save the current node in case cleaning the data triggers a
+        // destruction of the current decoration. Avoid crashing when trying to
+        // access to the node.
+        auto vnode = node;
+        vnode->view->set_decoration(nullptr); // ViewDecoration dies here.
+    };
 
-        // TODO: remove this code when
-        // https://github.com/WayfireWM/wayfire/pull/1187 gets released.
-        // This shouldn't be done here but in wayfire itself.
-        node->refresh_geometry();
+    wf::signal_connection_t on_fullscreen = [&](wf::signal_data_t *) {
+        if (node->view->fullscreen) {
+            if (!is_hidden())
+                detach_surface();
+        } else {
+            if (is_hidden())
+                attach_surface();
+        }
+        node->view->damage();
     };
 
+    // Attach the decoration surface to the node;
+    void attach_surface();
+
+    // Detach the decoration surface from the node;
+    void detach_surface();
+
   public:
     ViewDecoration(ViewNodeRef node, nonstd::observer_ptr<Options> options)
         : node(node), options(options) {
 
         node->connect_signal("prefered-split-type-changed",
                              &on_prefered_split_type_changed);
         node->connect_signal("detached", &on_detached);
+        node->view->connect_signal("fullscreen", &on_fullscreen);
 
         const auto output = node->get_ws()->output;
         output->connect_signal("swf-deco-fini", &on_detached);
         output->connect_signal("swf-deco-config-changed", &on_config_changed);
+
+        auto surface_unique_ptr =
+            std::make_unique<DecorationSurface>(node, options);
+        surface_ref = surface_unique_ptr.get();
+        surface = std::move(surface_unique_ptr);
+        if (!node->view->fullscreen)
+            attach_surface();
     }
 
     ~ViewDecoration() override {
+        detach_surface();
```

Comment:
To prevent an assertion error:
```suggestion
        if (!is_hidden())
            detach_surface();
```